tests/unit/test_symmetric_crypto.cpp

```diff
@@ -0,0 +1,312 @@
+            _expectedData.insert(std::end(_expectedData), std::begin(block), std::end(block));
+        }
+    }
+
+    QueueOfVectorsMemoryStrategy sut;
+
+    const int kNumberOfTestBlocks = 4;
+    const int kTestBlockLength = 16;
+    std::vector<uint8_t > _expectedData;
+};
+
+TEST_F(SymmetricCipherMemoryStrategy, SeveralWritesAndReadAll) {
+    sut.write({std::begin(_expectedData), std::begin(_expectedData) + 3});
+    sut.write({std::begin(_expectedData) + 3, std::begin(_expectedData) + 7});
+    sut.write({std::begin(_expectedData) + 7, std::end(_expectedData)});
+
+    auto readChunk = sut.readAll();
+    ASSERT_THAT(readChunk, ::testing::ElementsAreArray(_expectedData));
+}
+
+TEST_F(SymmetricCipherMemoryStrategy, EmptyReadAll) {
+    auto readBlock = sut.read(kTestBlockLength);
+    ASSERT_TRUE(readBlock.empty());
+}
+
+TEST_F(SymmetricCipherMemoryStrategy, ReadInWrittenBlockSizes) {
+    auto dataPacketizer = std::begin(_expectedData);
+
+    sut.write({dataPacketizer, dataPacketizer+=kTestBlockLength});
+    sut.write({dataPacketizer, dataPacketizer+=kTestBlockLength});
+    sut.write({dataPacketizer, dataPacketizer+=kTestBlockLength});
+
+    auto readBlock1 = sut.read(kTestBlockLength);
+    auto readBlock2 = sut.read(kTestBlockLength);
+
+    sut.write({dataPacketizer, dataPacketizer+=kTestBlockLength});
+
+    auto readBlock3 = sut.read(kTestBlockLength);
+    auto readBlock4 = sut.read(kTestBlockLength);
+
+    std::vector<uint8_t > readData;
+    readData.insert(std::end(readData), std::begin(readBlock1), std::end(readBlock1));
+    readData.insert(std::end(readData), std::begin(readBlock2), std::end(readBlock2));
+    readData.insert(std::end(readData), std::begin(readBlock3), std::end(readBlock3));
+    readData.insert(std::end(readData), std::begin(readBlock4), std::end(readBlock4));
+    ASSERT_THAT(readData, ::testing::ElementsAreArray(_expectedData));
+}
+
+TEST_F(SymmetricCipherMemoryStrategy, ReadInUnalignedBlockSizes) {
+    auto dataPacketizer = std::begin(_expectedData);
+    sut.write({dataPacketizer, dataPacketizer+=kTestBlockLength});
+    sut.write({dataPacketizer, dataPacketizer+=kTestBlockLength});
+    sut.write({dataPacketizer, dataPacketizer+=kTestBlockLength});
+
+    auto readBlock1 = sut.read(kTestBlockLength / 2);
+    auto readBlock2 = sut.read(kTestBlockLength * 1.6);
+
+    sut.write({dataPacketizer, dataPacketizer+=kTestBlockLength});
+
+    auto readBlock3 = sut.read(kTestBlockLength);
+    auto readBlock4 = sut.readAll();
+
+    std::vector<uint8_t > readData;
+    readData.insert(std::end(readData), std::begin(readBlock1), std::end(readBlock1));
+    readData.insert(std::end(readData), std::begin(readBlock2), std::end(readBlock2));
+    readData.insert(std::end(readData), std::begin(readBlock3), std::end(readBlock3));
+    readData.insert(std::end(readData), std::begin(readBlock4), std::end(readBlock4));
+    ASSERT_THAT(readData, ::testing::ElementsAreArray(_expectedData));
+}
```

Comment:
Missing: 
- Actual E2E Tests for the encryption. In particular, a given plaintext/ciphertext pair and see if this can be reproduced by our encryption/decryption code
- Test of the authentication mode: In particular, does it throw if there is a modification detected.